vllm/entrypoints/openai/serving_chat.py

```diff
@@ -318,20 +363,47 @@ async def chat_completion_stream_generator(
                         logprobs = None
 
                     delta_text = output.text[len(previous_texts[i]):]
-                    previous_texts[i] = output.text
-                    previous_num_tokens[i] = len(output.token_ids)
+                    delta_message: Optional[DeltaMessage] = None
 
-                    if request.tool_choice and type(
-                            request.tool_choice
-                    ) is ChatCompletionNamedToolChoiceParam:
+                    # handle streaming deltas for tools with tool_choice
+                    if (request.tool_choice and type(request.tool_choice) is
+                            ChatCompletionNamedToolChoiceParam):
                         delta_message = DeltaMessage(tool_calls=[
-                            ToolCall(function=FunctionCall(
+                            DeltaToolCall(function=DeltaFunctionCall(
                                 name=request.tool_choice.function.name,
-                                arguments=delta_text))
+                                arguments=delta_text),
+                                          index=i)
                         ])
+
+                    # handle streaming deltas for tools with tool_choice
+                    elif (request.tools and tool_parser
+                          and (request.tool_choice is None
+                               or request.tool_choice == 'auto')
+                          and self.enable_auto_tools):
```

Comment:
I've seen some conditions like this used a few places, worth refactoring to a helper check function if possible

Reply:
working on this in several places now :)